sdks/java/core/src/main/java/org/apache/beam/sdk/values/PCollectionViews.java

```diff
@@ -278,6 +528,398 @@ public IterableViewFn(TypeDescriptorSupplier<T> typeDescriptorSupplier) {
+          rangesWithSameFrom.add(
+              new OffsetRange(current.getFrom(), rangesToProcess.get(i).getTo()));
+        }
+
+        rangesToProcess.clear();
+      }
+      rangesWithSameFrom.add(current);
+    }
+
+    // Process the last chunk of overlapping ranges
+    while (!rangesWithSameFrom.isEmpty()) {
+      // This range always represents the range with with the smallest 'to'
+      OffsetRange current = rangesWithSameFrom.remove();
+
+      rangesToProcess.addAll(rangesWithSameFrom);
+      Collections.sort(rangesToProcess, OffsetRangeComparator.INSTANCE);
+      rangesWithSameFrom.clear();
+
+      rval.put(current, rangesToProcess.size() + 1 /* include current */);
+
+      // Shorten all the remaining ranges such that they start with current.to
+      for (OffsetRange rangeWithDifferentFrom : rangesToProcess) {
+        // Skip any duplicates of current
+        if (rangeWithDifferentFrom.getTo() > current.getTo()) {
+          rangesWithSameFrom.add(new OffsetRange(current.getTo(), rangeWithDifferentFrom.getTo()));
+        }
+      }
+      rangesToProcess.clear();
+    }
+    return rval.build();
+  }
+
+  @VisibleForTesting
+  static int computeTotalNumElements(
+      Map<OffsetRange, Integer> nonOverlappingRangesToNumElementsPerPosition) {
+    long sum = 0;
+    for (Map.Entry<OffsetRange, Integer> range :
+        nonOverlappingRangesToNumElementsPerPosition.entrySet()) {
+      sum +=
+          Math.multiplyExact(
+              Math.subtractExact(range.getKey().getTo(), range.getKey().getFrom()),
+              range.getValue());
+    }
+    return Ints.checkedCast(sum);
+  }
+
+  @VisibleForTesting
+  static KV<Long, Integer> computePositionForIndex(
+      Map<OffsetRange, Integer> nonOverlappingRangesToNumElementsPerPosition, int index) {
+    if (index < 0) {
+      throw new IndexOutOfBoundsException(
+          String.format(
+              "Position %s was out of bounds for ranges %s.",
+              index, nonOverlappingRangesToNumElementsPerPosition));
+    }
+    for (Map.Entry<OffsetRange, Integer> range :
+        nonOverlappingRangesToNumElementsPerPosition.entrySet()) {
+      int numElementsInRange =
+          Ints.checkedCast(
+              Math.multiplyExact(
+                  Math.subtractExact(range.getKey().getTo(), range.getKey().getFrom()),
+                  range.getValue()));
+      if (numElementsInRange <= index) {
+        index -= numElementsInRange;
+        continue;
+      }
+      long position = range.getKey().getFrom() + index / range.getValue();
+      int subPosition = index % range.getValue();
+      return KV.of(position, subPosition);
+    }
+    throw new IndexOutOfBoundsException(
+        String.format(
+            "Position %s was out of bounds for ranges %s.",
+            index, nonOverlappingRangesToNumElementsPerPosition));
+  }
+
+  /** Stores values or metadata about values. */
+  public static class MetaOr<T, MetaT> {
```

Comment:
Nit: Can you rename this class? Maybe something like "ValueOrMetadata"?

Reply:
Done